File: daily-coding-problem/dcp-jvm/dcp-scala/src/main/scala/dcp/problems/_1_20/Problem8_UnivalSubtreeCount.scala
package dcp.problems._1_20

/**
 * This problem was asked by Google.
 *
 * A unival tree (which stands for "universal value") is a tree where all nodes under it have the same value.
 *
 * Given the root to a binary tree, count the number of unival subtrees.
 */

object Problem8_UnivalSubtreeCount {

  case class Tree[A](data: A, left: Tree[A] = null, right: Tree[A] = null)

  type CountFlag = (Int, Boolean)

  def solution[A](root: Tree[A]): Int =
    if (root == null) 0
    else {
      val (count, _) = calcUni(root)
      count
    }

  def calcUni[A](root: Tree[A]): CountFlag = {
    if (root == null) return (0, true)
    val (leftCnt, isLeftUni) = calcUni(root.left)
    val (rightCnt, isRightUni) = calcUni(root.right)
    val totalCnt = leftCnt + rightCnt
    if (isLeftUni && isRightUni) {
      if (root.left != null && root.data != root.left.data) (totalCnt, false)
      else if (root.right != null && root.data != root.right.data) (totalCnt, false)
      else (totalCnt + 1, true)
    }
    else (totalCnt, false)
  }

}

File: daily-coding-problem/dcp-py/1_20/problem8_unival_subtree_count.py
"""
This problem was asked by Google.

A unival tree (which stands for "universal value") is a tree where all nodes under it have the same value.

Given the root to a binary tree, count the number of unival subtrees.

For example, the following tree has 5 unival subtrees:
```text
   0
  / \
 1   0
    / \
   1   0
  / \
 1   1
```
"""


class Tree:

    def __init__(self, data, left=None, right=None):
        self.data = data
        self.left = left
        self.right = right


def unival_subtree_count(tree):
    if not tree:
        return 0
    else:
        count, _ = calc_uni(tree)
        return count


def calc_uni(tree):
    if not tree:
        return 0, True

    left_cnt, is_left_uni = calc_uni(tree.left)
    right_cnt, is_right_uni = calc_uni(tree.right)
    total_cnt = left_cnt + right_cnt
    if is_left_uni and is_right_uni:
        if tree.left and tree.data != tree.left.data:
            return total_cnt, False
        elif tree.right and tree.data != tree.right.data:
            return total_cnt, False
        else:
            return total_cnt + 1, True
    else:
        return total_cnt, False


tree1 = Tree(5,
             left=Tree(1,
                       left=Tree(5),
                       right=Tree(5)),
             right=Tree(5,
                        right=Tree(5)))

assert unival_subtree_count(tree1) == 4

tree2 = Tree(0,
             left=Tree(1),
             right=Tree(0,
                        left=Tree(1,
                                  left=Tree(1),
                                  right=Tree(1)),
                        right=Tree(0)))
assert unival_subtree_count(tree2) == 5
